model/map/src/main/java/org/keycloak/models/map/realm/MapRealmProvider.java

```diff
@@ -0,0 +1,469 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.map.realm;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import org.jboss.logging.Logger;
+import static org.keycloak.common.util.StackUtil.getShortStackTrace;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.ClientScopeModel;
+import org.keycloak.models.GroupModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.ModelDuplicateException;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.RealmModel.SearchableFields;
+import org.keycloak.models.RealmProvider;
+import org.keycloak.models.RoleModel;
+import org.keycloak.models.map.common.Serialization;
+import org.keycloak.models.map.common.StreamUtils;
+import org.keycloak.models.map.common.StreamUtils.Pair;
+import org.keycloak.models.map.storage.MapKeycloakTransaction;
+import org.keycloak.models.map.storage.MapStorage;
+import org.keycloak.models.map.storage.ModelCriteriaBuilder;
+import org.keycloak.models.map.storage.ModelCriteriaBuilder.Operator;
+import org.keycloak.models.utils.KeycloakModelUtils;
+
+public class MapRealmProvider implements RealmProvider {
+
+    private static final Logger LOG = Logger.getLogger(MapRealmProvider.class);
+    private final KeycloakSession session;
+    final MapKeycloakTransaction<String, MapRealmEntity, RealmModel> tx;
+    private final MapStorage<String, MapRealmEntity, RealmModel> realmStore;
+
+    public MapRealmProvider(KeycloakSession session, MapStorage<String, MapRealmEntity, RealmModel> realmStore) {
+        this.session = session;
+        this.realmStore = realmStore;
+        this.tx = new MapKeycloakTransaction<>(realmStore);
+        session.getTransactionManager().enlist(tx);
+    }
+
+    private Function<MapRealmEntity, RealmModel> entityToAdapterFunc() {
+        // Clone entity before returning back, to avoid giving away a reference to the live object to the caller
+
+        return origEntity -> new MapRealmAdapter(session, registerEntityForChanges(origEntity));
```

Comment:
This doesn't need to be lambda since there is not realmId as the parameter. This method can be changed to `entityToAdapter`, that takes MapRealmEntity and returns RealmModel. Then we would have method reference instead of calling method that creates function.